documentation/modules/oauth/con-oauth-authorization-keycloak-example.adoc

```diff
@@ -0,0 +1,461 @@
+// Module included in the following module:
+//
+// con-oauth-authorization-keycloak-authorization-services.adoc
+
+[id='con-oauth-authorization-keycloak-example_{context}']
+= Example authorization rules configuration using Authorization Services
+
+This is an end-to-end example of using Keycloak Authorization Services to configure authorization rules for use with `keycloak` authorization in Strimzi.
+The example starts by deploying the Keycloak server with pre-configured realms, requiring no additional configuration.
+We deploy the Kafka cluster configured to use one of the pre-configured realms.
+We then connect to the Keycloak Admin Console and demonstrate the sections of the user interface where authorizations rules are configured.
+We use Kafka CLI client tools with different personal and service accounts to demonstrate limited access based on the permissions granted to different accounts.
+
+
+== Token-based authorization with Keycloak Authorization Services
+
+Once the Kafka Broker has obtained an access token by using `oauth` authentication, it is possible to use centrally managed authorization rules to enforce access restrictions onto Kafka Clients.
+For this, Strimzi Kafka Operator comes with `keycloak` authorization which uses Keycloak Authorization Services to centrally manage permissions.
+
+When using `keycloak` authorization, a custom authorizer is configured on the Kafka broker that uses Authorization Services REST endpoints available on Keycloak, which provide a list of granted permissions on resources for authenticated users.
+The list of grants (permissions) is fetched as the first action after an authenticated session is established by the Kafka client, and then regularly refreshed in the background.
+Grants are cached and enforced locally on the Kafka broker for each user session to provide fast authorization decisions. Because they are refreshed, any changes to the grants on the Keycloak server are detected and enforced.
+
+
+== Starting the pods
```

Comment:
I'm not sure we want to have this in the documentation. The installation should just link to Keycloak docs. That makes it clear that it is their responsibility. I think this docs should just start from the assumption that the user installed Keycloak with whatever their preferred method is.

And even if others think it should be here, it should definitely:
* not point to master but needs to point to a tag with the release used in given Strimzi release.
* Have a clear warning that this is not any secure production deployment.

Reply:
This is an example of integration of Kafka with Keycloak, and the `yaml`s are specifically configured to automatically import example realms that the example needs in order to work. These `yaml`s are part of `strimzi-kafka-oauth` examples. If you have no ready-made way to set up the example as it was intended, then this whole chapter is quite useless. And Strimzi example is not a concern of the Keycloak project. Also, the example chapter is really the reason for this PR. There is no Kubernetes based example anywhere else.

Pointing to a release tag is a great idea. Is there a mechanism in place that allows this, or is it something we have to develop?

We can also add a warning about this being just an insecure example, sure.

Reply:
> This is an example of integration of Kafka with Keycloak, and the yamls are specifically configured to automatically import example realms that the example needs in order to work.

I'm not sure this helps to be honest. Because you give them ready to use environment to play with it, but then you tell them that this is for nothing else then playing. So they would need to deploy it differently and then they would not know how to import the things?

> Pointing to a release tag is a great idea. Is there a mechanism in place that allows this, or is it something we have to develop?

There is variable `oauth-artifact-version` which should have the OAuth library version (which is the same as GitHub tag).

> We can also add a warning about this being just an insecure example, sure.

Not just insecure, but mainly not production ready. So I guess the warning should be for both.

Reply:
@mstruk Can you look at this so that we can merge this?

Reply:
There's several issues mentioned here:
- this content should in your opinion not be here - there should be a pointer to Keycloak documentation and then just instructions on how to import the realms. Problem: Importing the realm is only possible by mounting the realm file in yaml and deploying a new pod definition - it can not be done through UI on the existing running Keycloak instance because all the passwords, and secrets are lost in that case. But then the existing instructions do just that.
- need to use the oauth-artifact-version variable
- add warnings

Would addressing the last two issues be good enough?

Reply:
I guess I gave up on the first point I guess. WDYT @strimzi/maintainers?

The second and third point should be fixed.

Reply:
Added the variables, and a caution note.